enzy_htp/structure/structure.py

```diff
@@ -88,100 +90,189 @@
-        """Gets a chain of the given name. Returns None if the Chain() is not present."""
+        '''Gets a chain of the given name. Returns None if the Chain() is not present.'''
         return self.chain_mapper.get(chain_name, None)
 
+    #endregion
+
+    #region === Getter-Prop ===
     @property
+    def num_chains(self) -> int:
+        '''Returns the number of Chain() objects in the current Structure().'''
+        return len(self._chains)
+
+    @property 
     def residues(self) -> List[Residue]:
-        """Return a list of the residues in the Structure() object sorted by (chain_id, residue_id)"""
-        result = list()
-        for ch in self.chains:
-            result.extend(ch.residues())
-        result.sort(key=lambda r: r.sort_key())
+        '''Return a list of the residues in the Structure() object sorted by (chain_id, residue_id)'''
+        result = list(itertools.chain.from_iterable(self._chains))
+        result.sort(key=lambda r: r.key())
+        return result
+
+    @property
+    def ligands(self) -> List[Residue]:
+        '''Filters out the ligand Residue()'s from the chains in the Structure().'''
+        result: List[Residue] = list()
+        for chain in self.chains:
+            result.extend(list(filter(lambda r: r.is_ligand(), chain)))
         return result
+    #endregion
+
+    #region === Checker ===
+    def has_duplicate_chain_name(self) -> bool:
+        '''check if self._chain have duplicated chain name
+        give warning if do.'''
+        existing_c_id = []
+        ch: Chain
+        for ch in self._chains:
+            if ch.name in existing_c_id:
+                _LOGGER.warning(
+                    f'Duplicate chain names detected in Structure obj during {sys._getframe().f_back.f_code.co_name}()! '
+                )
+                return True
+            existing_c_id.append(ch.name)
+        return False
+    
+    def resolve_duplicated_chain_name(self) -> None:
+        '''resolve for duplicated chain name in self.chains_
+        A. insert the chain to be one character after the same one if they are different
+           and move the rest chain accordingly
+        B. give error if they are the same (in coordinate)'''        
+        mapper = {}
+        if_rename = 0
+        for ch in self.chains:
+            if ch.name in mapper:
+                if ch.is_same_coord(mapper[ch.name]):
+                    _LOGGER.error(
+                        'Duplicate chain (same coordinate) detected in Structure obj! Exiting... '
+                    )
+                    sys.exit(1)
+                new_name = chr(ord(ch.name) + 1) # TODO find a way
+                ch.name = new_name
+                if_rename = 1
+            mapper[ch.name] = ch
+        if if_rename:
+            _LOGGER.warning(
+                'Resolved duplicated chain (different ones) name by renaming.'
+            )
+    #endregion
+
+    #region === Editor ===
```

Comment:
I have noticed these `#region === Editor ===` and `#endregion` comments throughout. Please remove these.

Reply:
These are vscode region folding feature that helps organizing the code.